src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,282 @@
+  React.useEffect(() => {
+    setSelectedValue(value);
+  }, [value]);
+
+  // Fire onChange callback every time selected value changes
+  React.useEffect(() => {
+    onChange?.(selectedValue ?? null);
+  }, [selectedValue, onChange]);
+
+  // Controlled input value
+  const [inputValue, setInputValue] = React.useState<string>('');
+  const onInput = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setInputValue(event.target.value);
+      inputProps?.onChange?.(event);
+    },
+    [inputProps],
+  );
+
+  // Update input value when option is selected
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    selectedOption && setInputValue(selectedOption.label);
+  }, [selectedValue, options]);
+
+  // Filter options when input value changes
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    if (
+      inputValue &&
+      inputValue.length > 0 &&
+      isOpen &&
+      selectedOption?.label !== inputValue // don't filter list if option is selected
+    ) {
+      const _filteredOptions = options.filter((option) =>
+        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
+      );
+      setFilteredOptions(_filteredOptions);
+      if (!_filteredOptions.find(({ value }) => value === selectedValue)) {
+        setSelectedValue(undefined);
+      }
+    } else {
+      setFilteredOptions(options);
+    }
+  }, [inputValue, options, selectedValue, isOpen]);
+
+  // Update focused value according to filtered list
+  React.useEffect(() => {
+    if (!filteredOptions.find(({ value }) => value === selectedValue)) {
+      setFocusedIndex(options.indexOf(filteredOptions[0]));
+    } else {
+      setFocusedIndex(
+        options.findIndex(({ value }) => value === selectedValue),
+      );
+    }
+  }, [filteredOptions, options, selectedValue]);
+
+  // Update input value and focused value when menu is closed
+  React.useEffect(() => {
+    if (!isOpen) {
+      const selectedIndex = options.findIndex(
+        ({ value }) => value === selectedValue,
+      );
+      if (selectedIndex > -1 && inputValue !== options[selectedIndex].label) {
+        setInputValue(options[selectedIndex].label);
+      }
+      setFocusedIndex(selectedIndex);
+    }
+  }, [inputValue, isOpen, options, selectedValue]);
+
+  const onKeyDown = React.useCallback(
+    (event: React.KeyboardEvent) => {
+      const focusedIndexInFilteredList = filteredOptions.findIndex(
+        (option) => option === options[focusedIndex],
+      );
+      switch (event.key) {
+        case 'ArrowDown':
+          const nextIndex = Math.min(
+            focusedIndexInFilteredList + 1,
+            filteredOptions.length - 1,
+          );
+          setFocusedIndex(options.indexOf(filteredOptions[nextIndex]));
+          !isOpen && setIsOpen(true); // reopen menu if closed when typing
```

Comment:
Why arrow up it is not needed?

Reply:
My initial thought was: you go "down" from the input into the first/selected item, but you can't go "up" from the input into the menu.

Reply:
Finally found the right source for defined behavior. Says it's optional so maybe I'll leave it out
![image](https://user-images.githubusercontent.com/9084735/135162624-8cba4410-d861-47fc-8340-68c1577c911c.png)
https://w3c.github.io/aria-practices/#combobox